### ISSUE TITLE
How do we define a Tokenstream?

### ISSUE THREAD

**Issue by RunDevelopment** [PrismJS/prism#2071]
For #2070 I had to look at the [tests](https://github.com/PrismJS/prism/blob/master/tests/testrunner-tests.js) of `TokenStreamTransformer.simplify`. And from these tests, a token stream is defined as:

```ts
type TokenStream = Array<string | Token | TokenStream>;

interface Token {
    type: string;
    content: string | Token | TokenStream;
    // some other properties
}
```

Yes, [nested token streams](https://github.com/PrismJS/prism/blob/8ea685b8ca9f73784832f901f07cab25821f2eb0/tests/testrunner-tests.js#L71).

But I always assumed that a token stream looks like this:

```ts
type TokenStream = Array<string | Token>;
// Token same as above
```

And I even added the condition that a token stream doesn't contain adjacent strings. I assumed this because the token stream produced by `tokenize` has these properties and because these properties are very useful.

Point is: I wrote code with these assumptions, so do I have to rewrite it?

What is the correct definition of a token stream?

/cc @Golmote @mAAdhaTTah

**Comment by RunDevelopment** [comment]
We might even want to narrow down the definition of `Token` to:

```ts
type TokenStream = Array<string | Token>;
interface Token {
    type: string;
    content: string | TokenStream;
    // some other properties
}
```

because that is what `Prism.tokenize` actually produces.

**Comment by mAAdhaTTah** [comment]
Yeah, in general, if we can bring the types into alignment with what is actually produced at runtime, I'm ok with tightening up the types. It would only mean that consuming code has been handling cases it doesn't have to.

**Comment by RunDevelopment** [comment]
Closed by #1782.
